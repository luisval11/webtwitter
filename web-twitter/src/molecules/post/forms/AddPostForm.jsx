import PropTypes from "prop-types";
import React, { useState } from "react";
import { Alert, Form } from "react-bootstrap";
import { ButtonStyled } from "../../../atoms/ButtonStyled";
import { useForm } from "../../../hooks/useForm";
import { InputStyled } from "../../input/InputStyled";
import "./style.css";

export const AddPostForm = ({ setPosts }) => {
	const [error, setError] = useState(false);
	const [formValues, handleInputChange, reset] = useForm({
		content: "",
	});

	const handleSubmit = (e) => {
		e.preventDefault();
		if (formValues.content.trim() === "") {
			setError(true);
		} else {
			setPosts((posts) => [
				{
					id: Date.now(),
					authorId: 0,
					content: formValues.content,
					createdDate: new Date(),
				},
				...posts,
			]);
			reset();
		}
	};

	return (
		<Form onSubmit={handleSubmit}>
			{error && (
				<Alert
					className="mt-2"
					variant="danger"
					onClose={() => setError(false)}
					dismissible
				>
					<Alert.Heading>An error on the form submitted</Alert.Heading>
					<p>To create a new post, add a message on the textarea</p>
				</Alert>
			)}

			<InputStyled
				role="textarea"
				as="textarea"
				className="input__textarea"
				handleInput={handleInputChange}
				label="Post a new message"
				name="content"
				value={formValues.content}
			/>
			<div className="form__button--submit">
				<ButtonStyled onClick={handleSubmit} text="Post" />
			</div>
		</Form>
	);
};

AddPostForm.propTypes = {
	setPosts: PropTypes.func.isRequired,
};

export default AddPostForm;
