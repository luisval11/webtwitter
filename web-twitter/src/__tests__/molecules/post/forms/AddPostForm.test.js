import { render, fireEvent } from "@testing-library/react";
import AddPostForm from "../../../../molecules/post/forms/AddPostForm";

describe("Test on AddPostForm", () => {
	test("should match initial snapshot for AddPostForm", () => {
		const setPosts = jest.fn();
		let wrapper = render(<AddPostForm setPosts={setPosts} />);
		expect(wrapper).toMatchSnapshot();
	});

	test("should have 1 textarea", () => {
		const setPosts = jest.fn();
		const { container } = render(<AddPostForm setPosts={setPosts} />);
		const inputLength = container.getElementsByClassName("form-control").length;
		expect(inputLength).toBe(1);
	});

	test("should have 1 button", () => {
		const setPosts = jest.fn();
		const { container } = render(<AddPostForm setPosts={setPosts} />);
		const buttonLength = container.getElementsByClassName("btn-primary").length;
		expect(buttonLength).toBe(1);
	});

	test("should reset the value of the 'content' input when button is clicked", () => {
		const setPosts = jest.fn();
		const { container } = render(<AddPostForm setPosts={setPosts} />);
		const button = container.getElementsByClassName("btn-primary")[0];
		const input = container.getElementsByClassName("form-control")[0];
		fireEvent.change(input, { target: { value: content } });
		fireEvent.click(button);
		expect(setPosts).toHaveBeenCalledTimes(1);
	});
});
