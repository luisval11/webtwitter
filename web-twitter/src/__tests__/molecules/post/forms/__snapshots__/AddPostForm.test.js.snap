// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on AddPostForm should match initial snapshot for AddPostForm 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class=""
      >
        <div
          class="input__textarea"
        >
          <label
            class="form-label"
            for="content"
          >
            Post a new message
          </label>
          <textarea
            class="form-control"
            id="content"
            role="textarea"
          />
        </div>
        <div
          class="form__button--submit"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Post
          </button>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class=""
    >
      <div
        class="input__textarea"
      >
        <label
          class="form-label"
          for="content"
        >
          Post a new message
        </label>
        <textarea
          class="form-control"
          id="content"
          role="textarea"
        />
      </div>
      <div
        class="form__button--submit"
      >
        <button
          class="btn btn-primary"
          type="button"
        >
          Post
        </button>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
